File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
import copy

def go(maps):
    n = len(maps[0])
    m = len(maps)
    adj = copy.deepcopy(maps)
    dy = [0,0,1,-1]
    dx = [1,-1,0,0]
    
    # 방문해야할 node 저장
    queue = deque([(0,0)])
    cnt = 1
    
    while queue:
        (cy,cx) = queue.popleft()
        for i in range(4):
            ny = cy + dy[i]
            nx = cx + dx[i]
            if 0 <= ny < m and 0 <= nx < n and adj[ny][nx] == 1:
                adj[ny][nx] = adj[cy][cx] + 1
                queue.append((ny,nx))
                
    return adj[m-1][n-1]
    
def solution(maps):
    ret = go(maps)
    return ret if ret > 1 else -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js

function solution(maps) {
    const dy = [0,0,1,-1];
    const dx = [1,-1,0,0];
    
    // y , x , dist
    let queue = [[0,0,1]];
    
    while(queue.length > 0){
        const [cy,cx, dist] = queue.shift()
        
        if(cy === maps.length-1 && cx === maps[0].length -1){
            return dist;
        }
        
        for(let i =0; i < 4; i++){
            const ny = cy + dy[i];
            const nx = cx + dx[i];
            
            // 유효한 좌표일 때
            if(ny >=0 && ny < maps.length && nx >=0 && nx < maps[0].length){
                // 방문하지 않은 경우에만
                if(maps[ny][nx] === 1){
                    queue.push([ny,nx, dist+1]);
                    // 방문 처리
                    maps[ny][nx]++;
                }
            }
        }
    }
    
    return -1
}